File: lib/components/react_component.ts
export class ReactComponent extends MPPlatformView {
  refId = Math.random().toString();

  elementType(): string {
    return "div";
  }

  reactClass(): any {
    return undefined;
  }

  withIntrinsicContentSize(): boolean {
    return false;
  }

  setAttributes(attributes: any) {
    super.setAttributes(attributes);
    this.updateRenderObject();
  }

  extraAttributes() {
    return {};
  }

  setChildren() {}

  static makeChildren(owner: ReactComponent, children: any[]): any[] {
    return children
      ?.map((it: any) => {
        if (it.type === "text") {
          return React.createElement("span", {}, it.text);
        } else if (it.type === "icon") {
          return React.createElement(icons[it.data]);
        } else if (it.type === "link") {
          return React.createElement(
            it.highlighted === true ? antd.Typography.Link : "span",
            {
              href: it.href,
              target: it.target,
              style: { cursor: "pointer" },
              onClick: it.onClick
                ? () => {
                    owner.invokeMethod("onClick", { target: it.onClick });
                  }
                : undefined,
            },
            this.makeChildren(owner, it.children)
          );
        } else if (it.type === "innerComponent") {
          try {
            let clazz = this.getComponentClass(it.componentName);
            return React.createElement(
              clazz,
              it.attributes,
              this.makeChildren(owner, it.children)
            );
          } catch (error) {}
        }
      })
      ?.filter((it) => it);
  }

  static getComponentClass(name: string) {
    let nameComponents = name.split(".");
    let cur: any = window;
    for (let index = 0; index < nameComponents.length; index++) {
      cur = cur[nameComponents[index]];
    }
    return cur;
  }

  waitingGlobalScopeReady() {
    return new Promise((res) => {
      if (
        !window.ReactDOM ||
        !window.React ||
        !window.icons ||
        !window.antd ||
        !window.moment
      ) {
      } else {
        res(true);
        return;
      }
      setTimeout(() => {
        if (
          !window.ReactDOM ||
          !window.React ||
          !window.antd ||
          !window.icons ||
          !window.moment
        ) {
          res(false);
        } else {
          res(true);
        }
      }, 200);
    });
  }

  async updateRenderObject() {
    while ((await this.waitingGlobalScopeReady()) !== true) {}
    const children = ReactComponent.makeChildren(
      this,
      this.attributes.children
    );
    ReactDOM.render(
      React.createElement(
        this.reactClass(),
        {
          id: "d_" + this.refId,
          ...this.attributes,
          ...this.extraAttributes(),
        },
        children?.length === 1 ? children[0] : children
      ),
      this.htmlElement
    );
    if (this.withIntrinsicContentSize()) {
      const measuringNode = this.htmlElement.cloneNode(true) as HTMLElement;
      measuringNode.style.width = "unset";
      measuringNode.style.height = "unset";
      measuringNode.style.minWidth = cssLength(
        this.attributes.layoutConstraints.minWidth
      );
      measuringNode.style.maxWidth = cssLength(
        this.attributes.layoutConstraints.maxWidth
      );
      measuringNode.style.minHeight = cssLength(
        this.attributes.layoutConstraints.minHeight
      );
      measuringNode.style.maxHeight = cssLength(
        this.attributes.layoutConstraints.maxHeight
      );
      const result = measureHtmlElement(measuringNode);
      this.setSize({ width: result.width, height: result.height });
    }
  }
}

const cssLength = (value: string) => {
  if (value === "Infinity") {
    return "max-content";
  } else {
    return parseFloat(value) + "px";
  }
};

const measureHtmlElement = (htmlElement: HTMLElement) => {
  htmlElement.style.position = "fixed";
  htmlElement.style.top = "0px";
  htmlElement.style.left = "0px";
  htmlElement.style.opacity = "0px";
  document.body.appendChild(htmlElement);
  const rect = htmlElement.getBoundingClientRect();
  htmlElement.remove();
  return {
    width: rect?.width ?? 0,
    height: rect?.height ?? 0,
  };
};

File: lib/components/react_component.dart
abstract class AntComponent {}

class AntText extends AntComponent {
  final String text;
  AntText({required this.text});

  Map toJson() {
    return {
      'type': 'text',
      'text': text,
    };
  }
}

class AntIcon extends AntComponent {
  final String data;
  AntIcon({required this.data});

  Map toJson() {
    return {
      'type': 'icon',
      'data': data,
    };
  }
}

class AntLink extends AntComponent {
  final List<AntComponent>? children;
  final bool? highlighted;
  final String? href;
  final String? target;
  final Function? onClick;

  AntLink({
    String? text,
    AntIcon? icon,
    List<AntComponent>? children,
    this.highlighted,
    this.href,
    this.target,
    this.onClick,
  }) : this.children = (() {
          if (children != null) return children;
          if (text != null && icon != null) return [icon, AntText(text: text)];
          if (text != null) return [AntText(text: text)];
          if (icon != null) return [icon];
        })();

  static void handleMethodCall(
    String methodName,
    dynamic arguments,
    List<AntComponent> children,
  ) {
    if (methodName == 'onClick' &&
        arguments is Map &&
        arguments['target'] != null) {
      final target = findChild(arguments['target'], children);
      if (target != null && target is AntLink) {
        target.onClick?.call();
      }
    }
  }

  static AntComponent? findChild(int target, List<AntComponent> children) {
    AntComponent? result;
    for (var item in children) {
      if (item.hashCode == target) {
        result = item;
        break;
      } else if (item is AntInnerComponent && item.children != null) {
        var subResult = findChild(target, item.children!);
        if (subResult != null) {
          result = subResult;
          break;
        }
      } else if (item is AntLink && item.children != null) {
        var subResult = findChild(target, item.children!);
        if (subResult != null) {
          result = subResult;
          break;
        }
      }
    }
    return result;
  }

  Map toJson() {
    return {
      'type': 'link',
      'children': children,
      'highlighted': highlighted,
      'href': href,
      'target': target,
      'onClick': onClick != null ? hashCode : null,
    };
  }
}

class AntInnerComponent extends AntComponent {
  final String componentName;
  final Map? attributes;
  final List<AntComponent>? children;

  AntInnerComponent({
    required this.componentName,
    this.attributes,
    this.children,
  });

  Map toJson() {
    return {
      'type': 'innerComponent',
      'componentName': componentName,
      'attributes': attributes?..removeWhere((key, value) => value == null),
      'children': children,
    };
  }
}
